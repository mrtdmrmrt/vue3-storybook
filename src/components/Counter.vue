<template>
  <p>{{ label }}</p>
  <!-- @slot Slot to show content below label -->
  <slot name="sub-label" />
  <div class="container" :class="variant">
    <button @click="increment()">+</button>
    <p class="value">{{ count }}</p>
    <button @click="decrement()">-</button>
  </div>
  <!-- @slot Default slot to show any content below the counter -->
  <slot />
</template>

<script lang="ts">
import { ref, watch, PropType } from 'vue'
import { Variant } from './types'
/**
 * This is my amazing counter component
 *
 * It can increment and decrement!
 */
export default {
  props: {
    /**
     * The initial value for the counter
     */
    initialValue: {
      type: Number as PropType<number>,
      default: 0,
    },
    /**
     * Text shown above the counter
     */
    label: {
      type: String as PropType<string>,
      default: 'Counter',
    },
    /**
     * If true, the counter can show negative numbers
     */
    allowNegativeValues: {
      type: Boolean as PropType<boolean>,
      default: false,
    },
    /**
     * Defines the visual appearance of the counter
     */
    variant: {
      type: String as PropType<Variant>,
      default: Variant.Default,
    },
  },
  emits: ['counter-update'],
  setup(props: any, context: any) {
    const count = ref(props.initialValue)
    const increment = () => {
      count.value += 1
    }
    const decrement = () => {
      const newValue = count.value - 1
      if (newValue < 0 && !props.allowNegativeValues) {
        count.value = 0
      } else {
        count.value -= 1
      }
    }
    watch(count, (value) => {
      context.emit('counter-update', value)
    })
    return {
      count,
      increment,
      decrement,
    }
  },
}
</script>
<style scoped>
.container {
  display: flex;
  justify-content: space-between;
  border: 1px solid gray;
  border-radius: 10px;
  padding: 10px 10px;
}
.container.colored {
  background-color: lightgoldenrodyellow;
}
button {
  min-width: 50px;
  font-size: 24px;
}
.value {
  font-weight: bold;
  font-size: 24px;
}
</style>