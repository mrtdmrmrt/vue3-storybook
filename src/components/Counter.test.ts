import { mount } from '@vue/test-utils'

import Counter from './Counter.vue'

//ðŸ‘‡ Imports a specific story for the test
import { Colored, Default } from './Counter.stories'

it('renders default button', () => {
  const wrapper = mount(Counter, {
    propsData: Default.args,
  })
  expect(wrapper.find('.container').classes()).toContain('default')
})

it('renders colored button', () => {
  const wrapper = mount(Counter, {
    propsData: Colored.args,
  })
  expect(wrapper.find('.container').classes()).toContain('colored')
})

it('renders snapshot', () => {
  const wrapper = mount(Counter, {
    propsData: Colored.args,
  })
  expect(wrapper.element).toMatchSnapshot()
})