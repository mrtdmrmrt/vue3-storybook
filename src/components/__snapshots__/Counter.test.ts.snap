// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders snapshot 1`] = `
<div
  data-v-app=""
>
  
  <p>
    Colored
  </p>
  <!-- @slot Slot to show content below label -->
  
  
  <div
    class="container colored"
  >
    <button>
      +
    </button>
    <p
      class="value"
    >
      0
    </p>
    <button>
      -
    </button>
  </div>
  <!-- @slot Default slot to show any content below the counter -->
  
  
  
</div>
`;
